This notebook contains the different models and their accuracy using BMPA model.

# DATA LOADING


In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('data_up.csv')

In [3]:
data.head(5)

SEQN  Gender  Age  Annual-Family-Income  Ratio-Family-Income-Poverty  \
0     2       1   77                     8                         5.00   
1     5       1   49                    11                         5.00   
2    12       1   37                    11                         4.93   
3    13       1   70                     3                         1.07   
4    14       1   81                     5                         2.67   

   X60-sec-pulse  Systolic  Diastolic  Weight  Height  ...  Total-Cholesterol  \
0             68        98         56    75.4   174.0  ...               5.56   
1             66       122         83    92.5   178.3  ...               7.21   
2             64       174         99    99.2   180.0  ...               4.03   
3            102       130         66    63.6   157.7  ...               8.12   
4             72       136         61    75.5   166.2  ...               4.50   

    HDL  Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
0  1.39              4.7              3              3                 1   
1  1.08              5.5              1              1                 1   
2  0.98              5.2              2              1                 1   
3  1.28              7.6              3              3                 1   
4  1.04              5.8              1              1                 1   

   Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  CoronaryHeartDisease  
0         2                   2                 2                     0  
1         2                   2                 2                     0  
2         2                   1                 1                     0  
3         1                   1                 2                     0  
4         2                   2                 2                     0  

[5 rows x 51 columns]

In [4]:
data.columns

Index(['SEQN', 'Gender', 'Age', 'Annual-Family-Income',
       'Ratio-Family-Income-Poverty', 'X60-sec-pulse', 'Systolic', 'Diastolic',
       'Weight', 'Height', 'Body-Mass-Index', 'White-Blood-Cells',
       'Lymphocyte', 'Monocyte', 'Eosinophils', 'Basophils', 'Red-Blood-Cells',
       'Hemoglobin', 'Mean-Cell-Vol', 'Mean-Cell-Hgb-Conc.',
       'Mean-cell-Hemoglobin', 'Platelet-count', 'Mean-Platelet-Vol',
       'Segmented-Neutrophils', 'Hematocrit', 'Red-Cell-Distribution-Width',
       'Albumin', 'ALP', 'AST', 'ALT', 'Cholesterol', 'Creatinine', 'Glucose',
       'GGT', 'Iron', 'LDH', 'Phosphorus', 'Bilirubin', 'Protein', 'Uric.Acid',
       'Triglycerides', 'Total-Cholesterol', 'HDL', 'Glycohemoglobin',
       'Vigorous-work', 'Moderate-work', 'Health-Insurance', 'Diabetes',
       'Blood-Rel-Diabetes', 'Blood-Rel-Stroke', 'CoronaryHeartDisease'],
      dtype='object')

In [5]:
data.shape

(37079, 51)

In [6]:
data.isnull().sum()

SEQN                           0
Gender                         0
Age                            0
Annual-Family-Income           0
Ratio-Family-Income-Poverty    0
X60-sec-pulse                  0
Systolic                       0
Diastolic                      0
Weight                         0
Height                         0
Body-Mass-Index                0
White-Blood-Cells              0
Lymphocyte                     0
Monocyte                       0
Eosinophils                    0
Basophils                      0
Red-Blood-Cells                0
Hemoglobin                     0
Mean-Cell-Vol                  0
Mean-Cell-Hgb-Conc.            0
Mean-cell-Hemoglobin           0
Platelet-count                 0
Mean-Platelet-Vol              0
Segmented-Neutrophils          0
Hematocrit                     0
Red-Cell-Distribution-Width    0
Albumin                        0
ALP                            0
AST                            0
ALT                            0
Cholesterol                    0
Creatinine                     0
Glucose                        0
GGT                            0
Iron                           0
LDH                            0
Phosphorus                     0
Bilirubin                      0
Protein                        0
Uric.Acid                      0
Triglycerides                  0
Total-Cholesterol              0
HDL                            0
Glycohemoglobin                0
Vigorous-work                  0
Moderate-work                  0
Health-Insurance               0
Diabetes                       0
Blood-Rel-Diabetes             0
Blood-Rel-Stroke               0
CoronaryHeartDisease           0
dtype: int64

Conclusion : Data is clean and all the data is numerical
CoronaryHeartDisease is target class. This is a classification problem predicting the heart disease.
SEQN is just representing the row number hence we need not take in the prediction columns.
Here, we have 49 features and 1 target column.
We have originally 37079 rows with no cell being empty.

# DATA ANALYSIS


In [7]:
data.describe()

SEQN        Gender           Age  Annual-Family-Income  \
count  37079.000000  37079.000000  37079.000000          37079.000000   
mean   48901.041236      1.513282     48.943661              7.358208   
std    26753.636441      0.499830     18.010440              3.994083   
min        2.000000      1.000000     20.000000              1.000000   
25%    26120.500000      1.000000     33.000000              4.000000   
50%    50065.000000      2.000000     48.000000              7.000000   
75%    71173.500000      2.000000     63.000000             10.000000   
max    93702.000000      2.000000     85.000000             15.000000   

       Ratio-Family-Income-Poverty  X60-sec-pulse      Systolic     Diastolic  \
count                 37079.000000   37079.000000  37079.000000  37079.000000   
mean                      2.559026      72.579250    124.090078     69.919253   
std                       1.624789      12.242108     19.254741     13.575804   
min                       0.000000      32.000000      0.000000      0.000000   
25%                       1.140000      64.000000    111.000000     62.000000   
50%                       2.180000      72.000000    121.000000     70.000000   
75%                       4.130000      80.000000    134.000000     78.000000   
max                       5.000000     224.000000    270.000000    132.000000   

             Weight        Height  ...  Total-Cholesterol           HDL  \
count  37079.000000  37079.000000  ...       37079.000000  37079.000000   
mean      80.988276    167.389601  ...           5.081713      1.370344   
std       20.678734     10.122908  ...           1.072682      0.415985   
min       32.300000    129.700000  ...           1.530000      0.160000   
25%       66.500000    160.000000  ...           4.320000      1.070000   
50%       78.200000    167.100000  ...           5.020000      1.290000   
75%       92.100000    174.600000  ...           5.740000      1.600000   
max      371.000000    204.500000  ...          14.090000      5.840000   

       Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
count     37079.000000   37079.000000   37079.000000      37079.000000   
mean          5.676496       1.783840       1.598856          1.218587   
std           1.050223       0.448324       0.511199          0.461102   
min           2.000000       1.000000       1.000000          1.000000   
25%           5.200000       2.000000       1.000000          1.000000   
50%           5.400000       2.000000       2.000000          1.000000   
75%           5.800000       2.000000       2.000000          1.000000   
max          18.800000       3.000000       3.000000          9.000000   

           Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  \
count  37079.000000        37079.000000      37079.000000   
mean       1.907333            1.549502          1.796165   
std        0.349674            0.497550          0.402853   
min        1.000000            1.000000          1.000000   
25%        2.000000            1.000000          2.000000   
50%        2.000000            2.000000          2.000000   
75%        2.000000            2.000000          2.000000   
max        3.000000            2.000000          2.000000   

       CoronaryHeartDisease  
count          37079.000000  
mean               0.040670  
std                0.197527  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                1.000000  

[8 rows x 51 columns]

# Feature Selection


**ALGORITHM : BMPA**

In [8]:
!ls -l


total 8184
-rw-r--r-- 1 root root 8374838 Oct 11 16:19 data_up.csv
drwxr-xr-x 1 root root    4096 Oct  9 13:36 sample_data


In [9]:
!pip install mealpy


In [10]:
import mealpy
import pkgutil
mods = [name for _, name, _ in pkgutil.walk_packages(mealpy.__path__, mealpy.__name__ + ".")]

print([m for m in mods if "mpa" in m.lower() or "bmpa" in m.lower()])


['mealpy.swarm_based.MPA']


In [ ]:
!pip install --upgrade git+https://github.com/thieu1995/mealpy.git scikit-learn pandas numpy -q

Imports


In [12]:
import pandas as pd, numpy as np
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from mealpy.swarm_based.MPA import OriginalMPA
from mealpy.utils.space import BinaryVar

Load dataset


In [13]:
df = pd.read_csv('data_up.csv')

In [14]:
X = df.drop(columns=['SEQN', 'CoronaryHeartDisease'])

In [15]:
y = df['CoronaryHeartDisease']

Here we are defining fitness function by (minimize 1- accuracy)
And also we Train/Test split



In [16]:
def fitness(sol):
    mask = sol > 0.5
    if not mask.any():
        return 1
    Xs = X.loc[:, mask]

    Xtr, Xte, ytr, yte = train_test_split(Xs, y, test_size=0.2, random_state=42)
    clf = RandomForestClassifier(random_state=42).fit(Xtr, ytr)

    return 1 - accuracy_score(yte, clf.predict(Xte))

Binary Variables for MPA


In [17]:
bounds = [BinaryVar() for _ in range(X.shape[1])]

Defining Optimization Problem




In [18]:
problem = {
    "obj_func": fitness,
    "bounds": bounds,
    "minmax": "min",
}

Initializing Model and Running Binary MPA

In [19]:
model = OriginalMPA(epoch=5, pop_size=20)
result = model.solve(problem)

INFO:mealpy.swarm_based.MPA.OriginalMPA:OriginalMPA(epoch=5, pop_size=20)
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 1, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 151.60031 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 2, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 164.32165 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 3, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 163.63328 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 4, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 152.08680 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 5, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 163.83490 seconds


Extracting best solution


In [20]:
best_position = result.solution

if hasattr(result, "solution_fitness"):
    best_fitness = result.solution_fitness
elif hasattr(result, "fitness_history"):
    best_fitness = result.fitness_history[-1]
else:
    best_fitness = None
    print("Could not find fitness attribute in result. Inspect result.__dict__")



Could not find fitness attribute in result. Inspect result.__dict__


Selected Features


In [21]:
selected_features = np.where(best_position > 0.05)[0]
X_selected = X.iloc[:, selected_features] if selected_features.size > 0 else pd.DataFrame()

Calculating the Actual Accuracy and printing the results

In [22]:
if X_selected.shape[1] > 0:
    clf_final = RandomForestClassifier(random_state=42)
    accuracy = np.mean(cross_val_score(clf_final, X_selected, y, cv=5))
else:
    accuracy = 0



In [23]:
print("Best Fitness (1 - Accuracy):", best_fitness)


Best Fitness (1 - Accuracy): None


In [24]:
print("Selected Features:", selected_features)


Selected Features: [ 2  7  9 10 11 12 14 17 18 19 20 21 25 26 27 28 30 31 33 34 36 37 38 39
 40 42 43 44 45 46 47]


In [25]:
print("Total Features Selected:", len(selected_features))


Total Features Selected: 31


In [26]:
print("Accuracy of Selected Features:", accuracy)

Accuracy of Selected Features: 0.9592761400147374


**ALGORITHM : EBPSO**

In [39]:
!pip install mealpy --quiet


In [40]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

from mealpy.swarm_based.PSO import C_PSO
from mealpy.utils.space import BinaryVar


In [41]:
df = pd.read_csv('data_up.csv')
X, y = df.iloc[:, :-1], df.iloc[:, -1]



In [42]:

X = pd.DataFrame(X)
y = np.array(y).ravel()

Defining Fitness Function and Train/Test split

In [43]:
def fitness(sol):
    mask = sol > 0.5
    if not mask.any():
        return 1
    Xs = X.loc[:, mask]

    Xtr, Xte, ytr, yte = train_test_split(Xs, y, test_size=0.2, random_state=42)
    clf = RandomForestClassifier(random_state=42).fit(Xtr, ytr)

    return 1 - accuracy_score(yte, clf.predict(Xte))


Defining Binary Variables

In [44]:
bounds = [BinaryVar() for _ in range(X.shape[1])]
problem = {
    "obj_func": fitness,
    "bounds": bounds,
    "minmax": "min",
}


Run

In [45]:
# model = C_PSO(epoch=50, pop_size=20)
# result = model.solve(problem)

INFO:mealpy.swarm_based.PSO.C_PSO:C_PSO(epoch=50, pop_size=20, c1=2.05, c2=2.05, w_min=0.4, w_max=0.9)


KeyboardInterrupt: 

Extracting Best Solution

In [ ]:
best_position = result.solution
best_fitness = getattr(result, "solution_fitness", None) or getattr(result, "fitness_history", [None])[-1]

In [ ]:
selected_features_EBPSO = np.where(best_position > 0.5)[0]
X_selected = X.iloc[:, selected_features_EBPSO] if selected_features_EBPSO.size > 0 else pd.DataFrame()

Calculating Final Accuracy

In [ ]:
if X_selected.shape[1] > 0:
    clf_final = RandomForestClassifier(random_state=42)
    accuracy = np.mean(cross_val_score(clf_final, X_selected, y, cv=5))
else:
    accuracy = 0


Printing Results

In [ ]:
print("Best Fitness (1 - Accuracy):", best_fitness)
print("Selected Features Indices:", selected_features_EBPSO)
print("Total Features Selected:", len(selected_features_EBPSO))
print("Accuracy of Selected Features:", accuracy)


# MODEL SELECTION


In [46]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB

models = {
    'RandomForestClassifier' : RandomForestClassifier(random_state=42),
    'LogisticRegression' : LogisticRegression(random_state=42),
    'SupportVectorMachine' : SVC(random_state=42),
    'K-NearestNeighbors' : KNeighborsClassifier(),
    'DecisionTree' : DecisionTreeClassifier(random_state=42),
    'GaussianNaiveBayes' : GaussianNB()
}

In [48]:
X = df.drop(columns=['SEQN', 'CoronaryHeartDisease'])
y = df['CoronaryHeartDisease']
filtered_X = X.iloc[:, selected_features]

In [49]:
filtered_X

Annual-Family-Income  Weight  Body-Mass-Index  White-Blood-Cells  \
0                         8    75.4            24.90                7.6   
1                        11    92.5            29.10                5.9   
2                        11    99.2            30.62               10.2   
3                         3    63.6            25.57               11.6   
4                         5    75.5            27.33                9.1   
...                     ...     ...              ...                ...   
37074                    10    39.2            21.00                7.1   
37075                     4    59.1            21.50                6.4   
37076                     7    71.7            31.00                4.7   
37077                     1    78.2            26.00                7.6   
37078                     7    58.3            21.40                9.7   

       Lymphocyte  Monocyte  Basophils  Mean-Cell-Vol  Mean-Cell-Hgb-Conc.  \
0            21.1       7.1        0.5           88.5                 29.7   
1            37.8       6.2        0.4           84.9                 28.3   
2            23.7       9.0        0.6           83.5                 27.8   
3            13.1       3.8        0.4           91.1                 30.3   
4            29.8       5.6        0.4           90.4                 31.3   
...           ...       ...        ...            ...                  ...   
37074        24.2       8.9        0.6           80.5                 26.7   
37075        21.0      12.7        0.9           88.0                 30.0   
37076        45.9      12.5        0.8           81.6                 26.0   
37077        26.4       9.2        0.6           92.7                 30.8   
37078        24.6       5.5        0.3           86.3                 28.0   

       Mean-cell-Hemoglobin  ...  Protein  Uric.Acid  Triglycerides  \
0                 33.600000  ...     72.0      362.8          1.298   
1                 33.300000  ...     73.0      404.5          3.850   
2                 33.300000  ...     72.0      339.0          1.581   
3                 33.300000  ...     66.0      410.4          3.635   
4                 34.500000  ...     79.0      368.8          0.756   
...                     ...  ...      ...        ...            ...   
37074             33.859817  ...     76.0      428.3          1.264   
37075             32.559644  ...     69.0      368.8          0.948   
37076             32.981852  ...     66.0      273.6          1.095   
37077             33.867187  ...     73.0      237.9          0.937   
37078             33.398379  ...     68.0      249.8          0.892   

       Total-Cholesterol  Glycohemoglobin  Vigorous-work  Moderate-work  \
0                   5.56              4.7              3              3   
1                   7.21              5.5              1              1   
2                   4.03              5.2              2              1   
3                   8.12              7.6              3              3   
4                   4.50              5.8              1              1   
...                  ...              ...            ...            ...   
37074               4.14              5.8              2              2   
37075               3.62              5.6              2              2   
37076               6.62              5.7              2              2   
37077               3.72              5.2              2              2   
37078               4.14              4.6              2              2   

       Health-Insurance  Diabetes  Blood-Rel-Diabetes  
0                     1         2                   2  
1                     1         2                   2  
2                     1         2                   1  
3                     1         1                   1  
4                     1         2                   2  
...                 ...       ...                 ...  
37074                 1

In [51]:
for name, model in models.items():
    X_train, X_test, y_train, y_test = train_test_split(filtered_X, y, test_size=0.2, random_state=42)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"{name} Accuracy: {accuracy}")

RandomForestClassifier Accuracy: 0.9583333333333334


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression Accuracy: 0.9580636461704423
SupportVectorMachine Accuracy: 0.9583333333333334
K-NearestNeighbors Accuracy: 0.9568500539374326
DecisionTree Accuracy: 0.9211165048543689
GaussianNaiveBayes Accuracy: 0.9087108953613808
